### ISSUE TITLE
Additional features for working with modes

### ISSUE THREAD

**Issue by jan-david-fischbach** [flaport/meow#6]
Hey @flaport,
I am a big fan of meow. Besides its EME capabilities, I think it makes quite a lot of sense to use it as an abstraction layer over different FDE algorithms, as switching between "simulation backends" can otherwise necessitate significant code rework (for the structure definition alone). Maybe even interesting for gdsfactory (@joamatab) (We could also build a femwell backend for meow). With that in mind, there is a couple of things I'd like to add:
- [x] Poynting flux: auxiliary properties to calculate the poynting flux in x,y and z.

$$ \vec{P} = \vec{E} \times \vec{H} $$

An implementation using `np.cross` is probably most sensible. Addressed in #7.

- [x] Effective mode area: another auxiliary measure I tend to use relatively frequently. It is not that well defined/ agreed on in literature but I would use the definition as in [lumerical's implementation](https://optics.ansys.com/hc/en-us/articles/360034396734-FDE-solver-analysis-Mode-List-and-Deck):

$$ A_{eff} = \frac{\left(\int|\vec{E}|^2dA\right)^2}{\int|\vec{E}|^4dA} $$

- [ ] Implementation of the field energies for irregular meshes: [At this moment](https://github.com/flaport/meow/blob/18f84a7004aef2b0f046b7aee6344c0b361fdb5e/meow/mode.py#L232) the modal energies are computed by summing over the energy densities, which for equidistant meshes leads to a value proportional to the actual modal energy.:
```python
def electric_energy(mode: Mode) -> float:
    """get the electric energy contained in a `Mode`"""
    return electric_energy_density(mode).sum()
```
The proportionality constant can be eliminated by normalization, as it is not dependent on the field distribution. This does however not hold for non-equidistant meshes, as finely meshed regions are over-emphasized compared to the more coarsely meshed regions. I would therefore suggest dividing by the mesh step in x and y before summing or even better do a proper integration over the energy densities e.g. using nested `np.trapz` evaluations.

- [ ] I would love to be able to also easily get the group index. That would however require multiple frequency points to do a finite difference differentiation on the effective index with frequency/wavelength, right?

**Comment by flaport** [comment]
Hey @Jan-David-Black ,

Glad you like the library!

A few commonts on each of your points:

- I accepted PR #7 .
- Sticking to the Lumerical definition is the smart choice. It still is a leading standard on its own.
- I agree that generalizing some of the formulas for non-cartesian or non-uniform meshes is a good idea.
- We could add a group index function to the mode object which runs another mode simulation with exactly the same settings but a slightly perturbed wavelength. If we document this behavior (i.e. that this calculation is not instant) it might be a good solution.

**Comment by jan-david-fischbach** [comment]
Tidy3D doesn't yet seem to support non-equidistant meshes: https://github.com/flexcompute/tidy3d/issues/652
Lumerical does however.

> We could add a group index function to the mode object which runs another mode simulation with exactly the same settings but a slightly perturbed wavelength. If we document this behavior (i.e. that this calculation is not instant) it might be a good solution.

Interesting idea. We already have the cell available, so it should be relatively easy to do. Lumerical directly returns the group index. So in that case we can just cache it and avoid the second calculation. When I have time I'll open a PR

**Comment by simbilod** [comment]
@Jan-David-Black we have all these things in femwell, and started trying to combine it with MEOW to do EME (https://github.com/HelgeGehring/femwell/blob/main/femwell/eme.py). Not quite there yet, however

**Comment by smartalecH** [comment]
> I would love to be able to also easily get the group index. That would however require multiple frequency points to do a finite difference differentiation on the effective index with frequency/wavelength, right?

You can actually compute the group index from the fields directly (a simple integral from Snyder and Love, or you can rederive it using the [Hellmann-Feynman theorem](https://en.m.wikipedia.org/wiki/Hellmann–Feynman_theorem)). No need to do any extra simulations by perturbing the wavelength etc.

**Comment by flaport** [comment]
Thanks @smartalecH , I'll try that once I find some time :)

**Comment by jan-david-fischbach** [comment]
@smartalecH, you are talking about this formula, right?

$$
\tilde{v}_{\mathrm{g}}=-2 \frac{\tilde{\beta}}{k} \frac{c}{\lambda^3} \int_{A_x} \Psi^2 \mathrm{~d} A / \int_{A_x} \Psi^2 \frac{\mathrm{d}}{\mathrm{d} \lambda}\left(\frac{n^2}{\lambda^2}\right) \mathrm{d} A .
$$

Which is equation 33-16 of Snyder and Love.

**Comment by smartalecH** [comment]
> you are talking about this formula, right?

Yep that's right. There's lots of ways you can massage it into various forms depending on what the problem looks like and what you are after.
